### PR TITLE
Add Qwen3-VL decoder export support

### PR DESCRIPTION
**Pull request by sinitame**
## Description

This PR is the second step in the VLM export proposal (see #68 ) and adds initial decoder export support for the Qwen3-VL family.

 Following the handler-layer introduced in the previous PR, this change introduces a Qwen3-VL architecture handler. For Qwen3-VL, the decoder needs more than input_ids and KV cache. It also depends on multimodal state such as visual embeddings, grid metadata, and RoPE related state. This PR wires that support through the handler path so the exporter can support an initial VLM architecture without hardcoding Qwen-specific behavior into the shared wrapper/export flow.

This is intentionally scoped to decoder export only as the image encode + decoder will require some iterations, and potentially some work on the tract inference engine.

(it is based on `task/llm-loader-handler-split`)

### CODE REVIEW

File: packages/llm/torch_to_nnef_llm/models/handlers/qwen3_vl.py [hunk 1]
```diff
@@ -0,0 +1,419 @@
...
+            )
+            position_ids = self._build_cached_position_ids(
+                rope_deltas=rope_deltas_current,
+                past_seq_len=past_seq_len,
+                seq_length=input_ids.shape[1],
+                batch_size=input_ids.shape[0],
+                device=input_ids.device,
+            )
+
+        self._prev_rope_deltas = getattr(hf_model.model, "rope_deltas", None)
```
Comment (JulienBalianSonos) [inline packages/llm/torch_to_nnef_llm/models/handlers/qwen3_vl.py]: Suggest moving these off the handler instance into state_context.state (already used for image_embeddings / video_embeddings in this same function):

```python3
# instead of `self._...` writes:
state_context.state["prev_rope_deltas"] = getattr(hf_model.model, "rope_deltas", None)
state_context.state["last_rope_deltas"] = rope_deltas_current.detach().clone()
# or more cleanly postpone it to StateContext instanciation

#  then in build_forward_outputs: replace the two `self._...` reads:
if rope_deltas is None:
     rope_deltas = state_context.state["last_rope_deltas"]
     if hasattr(model.model, "rope_deltas"): 
         model.model.rope_deltas = state_context.state["prev_rope_deltas"]
```

Then drop both attributes from `__init__`. Two reasons:
1. For stage 3, the encoder handler will compose with this one state stashed on the decoder instance leaks across that boundary, while state_context is naturally scoped to one forward.
2. Handler instances are registered globally (one per model_type), so a retry of `build_forward_inputs` (or any concurrent export) races on these fields and the `hf_model.model.rope_deltas` restore can put back the wrong value -- unlikely today but easy to avoid.

File: packages/llm/torch_to_nnef_llm/models/handlers/qwen3_vl.py [hunk 2]
```diff
@@ -0,0 +1,419 @@
...
+        token_mask: torch.Tensor,
+        features: torch.Tensor,
+    ) -> torch.Tensor:
+        if features.numel() == 0:
+            return inputs_embeds
+
+        batch_size, seq_length = token_mask.shape
+        token_counts = token_mask.to(torch.long).sum(dim=-1)
+        total_tokens = int(token_counts.sum().item())
+        if total_tokens == 0 or total_tokens != features.shape[0]:
```
Comment (JulienBalianSonos) [inline packages/llm/torch_to_nnef_llm/models/handlers/qwen3_vl.py]: Here I propose:
```python
if total_tokens == 0:
    return inputs_embeds
if total_tokens != features.shape[0]:
    raise ValueError(
          f"feature/slot count mismatch: got {features.shape[0]} feature(s) "
          f"for {total_tokens} placeholder slot(s) in input_ids"
    )
```
Because to my understanding `total_tokens != features.shape[0]`: is a real bug from the caller: N feature vectors but M slots, with M != N. Silently returning un-injected inputs_embeds means the model runs with the wrong inputs, and we produce silently-wrong logits.

File: packages/llm/tests/test_handlers.py [hunk 3]
```diff
@@ -0,0 +1,155 @@
...
+from torch_to_nnef_llm.models.handlers import (
+    ArchitectureHandler,
+    DefaultArchitectureHandler,
+    get_handler,
+    register_handler,
+)
+from torch_to_nnef_llm.models.handlers.openelm import OpenELMArchitectureHandler
+from torch_to_nnef_llm.models.handlers.phi import PhiArchitectureHandler
+
+
```
Comment (JulienBalianSonos) [inline packages/llm/tests/test_handlers.py]: Nothing in this file touches Qwen3VLArchitectureHandler itself. 

Could we add at least:

1. A FakeQwen3VLModel smoke test mirroring `test_phi_wrapper_uses_handler_prepared_model_inputs`, asserting `BaseCausal.forward` reaches the inner model with the expected inputs_embeds / attention_mask / position_ids shapes (and that state_context.state ends up populated).
2. A _inject_token_features shape-mismatch test (see [this comment](https://github.com/sonos/torch-to-nnef/pull/70/changes#r3186923325)). 
3. A one-liner asserting `Qwen3VLArchitectureHandler.get_auto_model_class(fake_transformers)` returns `Qwen3VLForConditionalGeneration`: it's a new hook from #67 and qwen3_vl is its first non-trivial user.
 
Also if you have some more time and the end-to-end export-and-run-on-tract path is still flaky, it's fine to add a test like test_llama.py with gate (1) with tract_check_io=False / --no-verify so the export path itself is exercised even if the round-trip parity check isn't ready yet.

File: packages/llm/torch_to_nnef_llm/models/base.py [hunk 4]
```diff
@@ -233,37 +233,34 @@ def wrapped(self, *args, **kwargs):
     return wrapped
 
 
-def update_forward_signature(self):
+def update_forward_signature(self, io_spec=None):
```
Comment (JulienBalianSonos) [inline packages/llm/torch_to_nnef_llm/models/base.py:243]: The wrapper's forward signature now gets rewritten twice: once in `BaseCausal.__init__` (l338) with `io_spec=None` (default input_ids + kv_cache_*), then again in the exporter (exporter.py:l450) once the handler's build_input_spec has resolved, which for Qwen3-VL adds five extra named inputs. The second call wins; the first is just a placeholder.

Fine in itself, but the _forward / __signature__ trickery makes this tempting to "simplify" later. A short docstring would save the next contributor an hour:
 
```python
def update_forward_signature(self, io_spec=None):
    """Rewrite `forward`'s signature so torch>=2.0 export captures named inputs.

     Called twice: at `__init__` time with a default LLM signature, then again
     by the exporter once the handler's IOSpec is resolved. The second call wins.
   """
```

File: packages/llm/torch_to_nnef_llm/models/handlers/qwen3_vl.py [hunk 5]
```diff
@@ -0,0 +1,419 @@
...
+            "vision_start_token_id",
+            config_helper.conf.image_token_id - 1,
+        )
+        image_token_id = config_helper.conf.image_token_id
+        video_token_id = config_helper.conf.video_token_id
+
+        input_ids.random_(0, vocab_size)
+        if vocab_size > 10:
+            input_ids[input_ids == image_token_id] = 1
+            input_ids[input_ids == video_token_id] = 2
```
Comment (JulienBalianSonos) [inline packages/llm/torch_to_nnef_llm/models/handlers/qwen3_vl.py:268]: The literals 1 and 2 look arbitrary on first read. I assume they are just "any token that isn't image_token_id / video_token_id" so the random fill doesn't accidentally trigger the vision injection paths during tracing. A one-line comment to that effect (or named constants) would make it clear.

**Review by JulienBalianSonos — CHANGES_REQUESTED** [review]
Nice second step on top of #67. The handler abstraction holds up well under a real VLM: the IOSpec/StateContext extension is generic, shared-code changes don't leak Qwen specifics, and the get_auto_model_class hook lands in the right place. The composition pattern in build_input_spec (subclass + append + split back out) is reusable enough that a LLaVA-style handler should fit the same shape. Thanks a lot.

Marking as Request Changes given the inline minor comments.

**Comment by JulienBalianSonos** [comment]
Hey @sinitame, hope you're doing well! 👋

I've been looking at this, and I'm really excited about the Qwen3-VL decoder export landing. It's a feature I'd love to see in t2n.

I know things get busy, so no pressure at all, but if you're short on time these days I'm happy to pick it up and do the final push to get it across the line. Just let me know what works best for you and whether there's anything specific still in flight on your end.

Thanks again for kicking this off! 🙏

**Comment by sinitame** [comment]
Hi @JulienBalianSonos  !
Yes, I've had limited bandwidth over the past couple of weeks. I'll get back to it over the weekend to finish the parts I've started, and I'd be happy to get your help polishing it afterward.

File: packages/llm/torch_to_nnef_llm/models/handlers/qwen3_vl.py (hunk 1 above)
Comment (sinitame) [reply packages/llm/torch_to_nnef_llm/models/handlers/qwen3_vl.py]: Done in [e1167c2](https://github.com/sonos/torch-to-nnef/pull/70/commits/e1167c29784430eb3f39b443f054810af7547bdf). I moved the temporary RoPE state into StateContext.state and removed the handler instance fields, so the restore path is scoped to a single forward.

File: packages/llm/torch_to_nnef_llm/models/handlers/qwen3_vl.py (hunk 2 above)
Comment (sinitame) [reply packages/llm/torch_to_nnef_llm/models/handlers/qwen3_vl.py]: Addressed in [e1167c2](https://github.com/sonos/torch-to-nnef/pull/70/commits/e1167c29784430eb3f39b443f054810af7547bdf).

File: packages/llm/tests/test_handlers.py (hunk 3 above)
Comment (sinitame) [reply packages/llm/tests/test_handlers.py]: Done in [8cde663](https://github.com/sonos/torch-to-nnef/pull/70/commits/8cde663a308fe78c5fe5bd8bda7f937f050a42ed). Added Qwen3-VL handler coverage with:
- a fake model smoke test asserting the wrapper reaches the model with expected inputs_embeds / attention_mask / position_ids shapes and populated handler state
- a shape-mismatch test for `_inject_token_features`
- a `get_auto_model_class` assertion for `Qwen3VLForConditionalGeneration`

File: packages/llm/torch_to_nnef_llm/models/base.py (hunk 4 above)
Comment (sinitame) [reply packages/llm/torch_to_nnef_llm/models/base.py:243]: Yep true, I made the change in [8cde663](https://github.com/sonos/torch-to-nnef/pull/70/commits/8cde663a308fe78c5fe5bd8bda7f937f050a42ed).

File: packages/llm/torch_to_nnef_llm/models/handlers/qwen3_vl.py (hunk 5 above)
Comment (sinitame) [reply packages/llm/torch_to_nnef_llm/models/handlers/qwen3_vl.py:268]: I've added the comment

File: packages/llm/torch_to_nnef_llm/loader.py [hunk 6]
```diff
@@ -573,7 +581,13 @@ def load_model(
             dir_path = find_subdir_with_filename_in(local_dir, "config.json")
             assert dir_path.is_dir(), dir_path
             assert_model_safetensors_exists(dir_path)
-            hf_model_causal = _from_pretrained(dir_path, **kwargs)
+            config = transformers.AutoConfig.from_pretrained(
+                dir_path, trust_remote_code=True
+            )
```
Comment (JulienBalianSonos) [inline packages/llm/torch_to_nnef_llm/loader.py:608]: Blocker:
The `trust_remote_code=True` should be replaced by: `trust_remote_code=trust_remote_code` (both this call and the local-dir one just above at line 600).
We recently added this flag for finer security control: the promise is that `--no-trust-remote-code` refuses to execute the custom Python a repo can ship (its auto_map config/modeling code).

Since this new `AutoConfig.from_pretrained` runs on the shared load path for every slug and local dir, hardcoding True silently defeats that opt-out for all models, not just Qwen3-VL.

File: packages/llm/torch_to_nnef_llm/loader.py [hunk 7]
```diff
@@ -582,7 +596,13 @@ def load_model(
         except (T2NErrorNotFoundFile, OSError):
             hf_model_causal = load_peft_model(local_dir, kwargs)
     elif hf_model_slug is not None:
-        hf_model_causal = _from_pretrained(hf_model_slug, **kwargs)
+        config = transformers.AutoConfig.from_pretrained(
+            hf_model_slug, trust_remote_code=True
```
Comment (JulienBalianSonos) [inline packages/llm/torch_to_nnef_llm/loader.py]: see the comment upper.

File: packages/llm/torch_to_nnef_llm/exporter.py [hunk 8]
```diff
@@ -232,10 +232,19 @@ def __init__(
             self.hf_model_causal.config.torchscript = False
             # only effective if config set tie_word_embeddings=True
             # tie_encoder_decoder=True
-            self.hf_model_causal.tie_weights()
+        self.hf_model_causal.tie_weights()
 
         self.model_infos = HFConfigHelper(self.hf_model_causal.config)
-
+        if self.model_infos.conf.model_type == "qwen3_vl":
```
Comment (JulienBalianSonos) [inline packages/llm/torch_to_nnef_llm/exporter.py]: Polish:
this `model_type == "qwen3_vl"` branch (plus the nested `model.model.language_model` probing) is the one spot where arch specifics land in the shared exporter, which the handler layer was meant to avoid. A `FORCE_ATTN_IMPL` class attr or a `handler.prepare_model_for_export(model)` hook would keep it with the handler. Happy to take this as a follow-up so it doesn't hold up the merge.

File: packages/llm/torch_to_nnef_llm/models/handlers/phi.py [hunk 9]
```diff
@@ -46,38 +44,43 @@ def build_forward_inputs(
...
     ) -> T.List[torch.Tensor]:
         hidden_states = model_outputs[0]
         logits = model.lm_head(hidden_states[:, -num_logits_to_keep:, :])
-        kvs = [
-            t
-            for kv in dyn_cache_to_legacy(model_inputs["past_key_values"])
-            for t in kv
-        ]
+        past_key_values = state_context.model_inputs["past_key_values"]
+        if hasattr(past_key_values, "to_legacy_cache"):
```
Comment (JulienBalianSonos) [inline packages/llm/torch_to_nnef_llm/models/handlers/phi.py]: Polish:
This looks like accidental change from the StateContext swap: the inlined block is exactly `dyn_cache_to_legacy` (base.py:71), which phi.py still has access to. Reverting to `dyn_cache_to_legacy(...)` here keeps it a one-liner and avoids leaving the helper orphaned (it now has no other callers).

**Review by JulienBalianSonos — CHANGES_REQUESTED** [review]
Thanks for the thorough new round, @sinitame, this is really close. 

Only blocker before merge is the trust_remote_code fix in loader.py (it defeats --no-trust-remote-code for every model).

The rest is non-blocking, and I'm happy to pick up the follow-ups. Great step toward VLM support.

File: packages/llm/torch_to_nnef_llm/loader.py (hunk 6 above)
Comment (sinitame) [reply packages/llm/torch_to_nnef_llm/loader.py:608]: Good point, should be fixed now

**Review by JulienBalianSonos — APPROVED** [review]
LGTM thank you so much for pushing it that far !
